src/Data/Bin/Bin0.agda

```diff
@@ -0,0 +1,686 @@
+    2 *' ((1+ ((1+ m) +' n)))        ≡⟨ cong (2*'_ ∘ 1+_ ∘ (_+' n))
+                                                          (+'-comm 1 m) ⟩
+    2 *' ((1+ ((m +' 1) +' n)))      ≡⟨ cong (2*'_ ∘ 1+_) (+'-assoc m 1 n) ⟩
+    2 *' ((1+ ((m +' (1+ n)))))      ≡⟨ cong 2*'_
+                                                (sym (+'-assoc 1 m (1+ n))) ⟩
+    2 *' ((1+ m) +' (1+ n))          ≡⟨ *'-lDistrib 2 (1+ m) (1+ n) ⟩
+    (2*' (1+ m)) +' (2*' (1+ n))     ≡⟨⟩
+    toℕ (2suc x) +' toℕ (2suc y)
+  ∎
+  where m = toℕ x;  n = toℕ y
+
+toℕ+homo (2suc x) (suc2* y) =
+  begin
+    toℕ ((2suc x) + (suc2* y))     ≡⟨⟩
+    toℕ (suc (2suc (x + y)))       ≡⟨ toℕ∘suc (2suc (x + y)) ⟩
+    1+ (toℕ (2suc (x + y)))        ≡⟨⟩
+    1+ (2*' (1+ (toℕ (x + y))))    ≡⟨ cong (1+_ ∘ 2*'_ ∘ 1+_) (toℕ+homo x y) ⟩
+    1+ (2*' (1+ (m +' n)))          ≡⟨ cong (1+_ ∘ 2*'_)
+                                                      (sym (+'-assoc 1 m n)) ⟩
+    1+ (2*' (1+m +' n))             ≡⟨ cong 1+_ (*'-lDistrib 2 1+m n) ⟩
+    1+ ((2*' 1+m) +' 2*' n)         ≡⟨ Of+ℕ-semigroup.x∙yz≈y∙xz +'-comm
+                                                       {1} {_} {2*' n} ⟩
+    (2*' 1+m) +' (1+ (2*' n))       ≡⟨⟩
+    toℕ (2suc x) +' toℕ (suc2* y)
+  ∎
+  where m = toℕ x;  n = toℕ y;  1+m = 1+ m
+
+
+toℕ+homo (suc2* x) (2suc y) =
+  begin
+    toℕ ((suc2* x) + (2suc y))      ≡⟨⟩
+    toℕ (suc (2suc (x + y)))        ≡⟨ toℕ∘suc (2suc (x + y)) ⟩
+    1+ (toℕ (2suc (x + y)))         ≡⟨⟩
+    1+ (2*' (1+ (toℕ (x + y))))     ≡⟨ cong (1+_ ∘ 2*'_ ∘ 1+_) (toℕ+homo x y)
+                                      ⟩
+    1+ (2*' (1+ (m +' n)))          ≡⟨ cong (1+_ ∘ 2*'_)
+                                           (Of+ℕ-semigroup.x∙yz≈y∙xz +'-comm
+                                                            {1} {m} {n})
+                                     ⟩
+    1+ (2*' (m +' 1+n))             ≡⟨ cong 1+_ (*'-lDistrib 2 m 1+n) ⟩
+    1+ (2*' m +' 2*' 1+n)           ≡⟨ sym (+'-assoc 1 (2*' m) (2*' 1+n)) ⟩
+    (1+ (2*' m)) +' (2*' (1+ n))    ≡⟨⟩
+    toℕ (suc2* x) +' toℕ (2suc y)
+  ∎
+  where m = toℕ x;  n = toℕ y;  1+n = 1+ n
+
+toℕ+homo (suc2* x) (suc2* y) =
+  begin
+    toℕ ((suc2* x) + (suc2* y))    ≡⟨⟩
+    toℕ (suc (suc2* (x + y)))      ≡⟨ toℕ∘suc (suc2* (x + y)) ⟩
+    1+ (toℕ (suc2* (x + y)))       ≡⟨⟩
+    1+ (1+ (2*' (toℕ (x + y))))    ≡⟨ cong (1+_ ∘ 1+_ ∘ 2*'_) (toℕ+homo x y) ⟩
+    1+ (1+ (2*' (m +' n)))          ≡⟨ cong (1+_ ∘ 1+_) (*'-lDistrib 2 m n) ⟩
+    1+ (1+ (2*' m +' 2*' n))        ≡⟨ cong 1+_ (sym (+'-assoc 1 (2*' m) _)) ⟩
+    1+ ((1+ (2*' m) +' (2*' n)))    ≡⟨ (Of+ℕ-semigroup.x∙yz≈y∙xz +'-comm
+                                                       {1} {1+ (2*' m)} {_}) ⟩
+    (1+ (2*' m)) +' (1+ (2*' n))    ≡⟨⟩
+    toℕ (suc2* x) +' toℕ (suc2* y)
+  ∎
+  where m = toℕ x;  n = toℕ y
+
+-----------------------------------------
+fromℕ :  (n : ℕ) → ∃ (\x → toℕ x ≡ n)    -- Mind:  it costs O(n)
```

Comment:
As per standard library conventions, this function should only return a Bin

Reply:
Changed to
```
ext-fromℕ :  (n : ℕ) → ∃ (\x → toℕ x ≡ n)
...
fromℕ : ℕ → Bin
fromℕ = proj₁ ∘ ext-fromℕ
```